File: apps/api/app/routers/tickets.py
"""Ticket inbox/detail/reply APIs."""

from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.deps import get_current_session, get_db, require_csrf_header, require_permission
from app.core.policies import POLICIES
from app.schemas.auth import UserSession
from app.schemas.ticketing import (
    TicketDetailResponse,
    TicketListItem,
    TicketListResponse,
    TicketPatchRequest,
    TicketReplyRequest,
    TicketComposeRequest,
    TicketSendResult,
    TicketNoteCreateRequest,
    TicketNoteRead,
    TicketLinkSurrogateRequest,
    TicketSendIdentityResponse,
)
from app.services import ticketing_service

router = APIRouter(
    prefix="/tickets",
    tags=["Tickets"],
    dependencies=[Depends(require_permission(POLICIES["tickets"].default))],
)


@router.get("", response_model=TicketListResponse)
def list_tickets(
    limit: Annotated[int, Query(ge=1, le=100)] = 50,
    cursor: str | None = None,
    status: str | None = None,
    priority: str | None = None,
    queue_id: UUID | None = None,
    assignee_user_id: UUID | None = None,
    surrogate_id: UUID | None = None,
    needs_review: bool | None = None,
    q: str | None = None,
    db: Session = Depends(get_db),
    session: UserSession = Depends(get_current_session),
) -> TicketListResponse:
    """List tickets with cursor pagination + inbox filters."""
    page = ticketing_service.list_tickets(
        db,
        org_id=session.org_id,
        limit=limit,
        cursor=cursor,
        status_filter=status,
        priority_filter=priority,
        queue_id=queue_id,
        assignee_user_id=assignee_user_id,
        surrogate_id=surrogate_id,
        needs_review=needs_review,
        q=q,
    )
    items = [
        TicketListItem(
            id=ticket.id,
            ticket_code=ticket.ticket_code,
            status=ticket.status.value if hasattr(ticket.status, "value") else str(ticket.status),
            priority=ticket.priority.value if hasattr(ticket.priority, "value") else str(ticket.priority),
            subject=ticket.subject,
            requester_email=ticket.requester_email,
            requester_name=ticket.requester_name,
            assignee_user_id=ticket.assignee_user_id,
            assignee_queue_id=ticket.assignee_queue_id,
            surrogate_id=ticket.surrogate_id,
            surrogate_link_status=ticket.surrogate_link_status.value
            if hasattr(ticket.surrogate_link_status, "value")
            else str(ticket.surrogate_link_status),
            first_message_at=ticket.first_message_at,
            last_message_at=ticket.last_message_at,
            last_activity_at=ticket.last_activity_at,
            created_at=ticket.created_at,
            updated_at=ticket.updated_at,
        )
        for ticket in page.items
    ]
    return TicketListResponse(items=items, next_cursor=page.next_cursor)


@router.get("/send-identities", response_model=TicketSendIdentityResponse)
def list_send_identities(
    db: Session = Depends(get_db),
    session: UserSession = Depends(require_permission(POLICIES["tickets"].actions["reply"])),
) -> TicketSendIdentityResponse:
    """List available Gmail sender identities for the current user."""
    identities = ticketing_service.list_send_identities(db, user_id=session.user_id)
    return TicketSendIdentityResponse(items=identities)


@router.get("/{ticket_id}", response_model=TicketDetailResponse)
def get_ticket_detail(
    ticket_id: UUID,
    db: Session = Depends(get_db),
    session: UserSession = Depends(get_current_session),
) -> TicketDetailResponse:
    """Return ticket detail timeline and metadata."""
    payload = ticketing_service.get_ticket_detail(
        db,
        org_id=session.org_id,
        ticket_id=ticket_id,
    )
    return TicketDetailResponse(**payload)


@router.patch(
    "/{ticket_id}",
    response_model=TicketListItem,
    dependencies=[Depends(require_csrf_header)],
)
def patch_ticket(
    ticket_id: UUID,
    data: TicketPatchRequest,
    db: Session = Depends(get_db),
    session: UserSession = Depends(require_permission(POLICIES["tickets"].actions["edit"])),
) -> TicketListItem:
    """Update status/priority/assignment."""
    ticket = ticketing_service.patch_ticket(
        db,
        org_id=session.org_id,
        actor_user_id=session.user_id,
        ticket_id=ticket_id,
        status_value=data.status,
        priority_value=data.priority,
        assignee_user_id=data.assignee_user_id,
        assignee_queue_id=data.assignee_queue_id,
    )
    return TicketListItem(
        id=ticket.id,
        ticket_code=ticket.ticket_code,
        status=ticket.status.value if hasattr(ticket.status, "value") else str(ticket.status),
        priority=ticket.priority.value if hasattr(ticket.priority, "value") else str(ticket.priority),
        subject=ticket.subject,
        requester_email=ticket.requester_email,
        requester_name=ticket.requester_name,
        assignee_user_id=ticket.assignee_user_id,
        assignee_queue_id=ticket.assignee_queue_id,
        surrogate_id=ticket.surrogate_id,
        surrogate_link_status=ticket.surrogate_link_status.value
        if hasattr(ticket.surrogate_link_status, "value")
        else str(ticket.surrogate_link_status),
        first_message_at=ticket.first_message_at,
        last_message_at=ticket.last_message_at,
        last_activity_at=ticket.last_activity_at,
        created_at=ticket.created_at,
        updated_at=ticket.updated_at,
    )


@router.post(
    "/{ticket_id}/reply",
    response_model=TicketSendResult,
    dependencies=[Depends(require_csrf_header)],
)
def reply_ticket(
    ticket_id: UUID,
    data: TicketReplyRequest,
    db: Session = Depends(get_db),
    session: UserSession = Depends(require_permission(POLICIES["tickets"].actions["reply"])),
) -> TicketSendResult:
    """Send threaded reply via Gmail."""
    payload = ticketing_service.reply_to_ticket(
        db,
        org_id=session.org_id,
        actor_user_id=session.user_id,
        ticket_id=ticket_id,
        to_emails=data.to_emails,
        cc_emails=data.cc_emails,
        subject=data.subject,
        body_text=data.body_text,
        body_html=data.body_html,
        idempotency_key=data.idempotency_key,
    )
    return TicketSendResult(**payload)


@router.post(
    "/compose",
    response_model=TicketSendResult,
    dependencies=[Depends(require_csrf_header)],
)
def compose_ticket(
    data: TicketComposeRequest,
    db: Session = Depends(get_db),
    session: UserSession = Depends(require_permission(POLICIES["tickets"].actions["reply"])),
) -> TicketSendResult:
    """Compose a new outbound ticket message."""
    payload = ticketing_service.compose_ticket(
        db,
        org_id=session.org_id,
        actor_user_id=session.user_id,
        to_emails=data.to_emails,
        cc_emails=data.cc_emails,
        subject=data.subject,
        body_text=data.body_text,
        body_html=data.body_html,
        surrogate_id=data.surrogate_id,
        queue_id=data.queue_id,
        idempotency_key=data.idempotency_key,
    )
    return TicketSendResult(**payload)


@router.post(
    "/{ticket_id}/notes",
    response_model=TicketNoteRead,
    dependencies=[Depends(require_csrf_header)],
)
def add_ticket_note(
    ticket_id: UUID,
    data: TicketNoteCreateRequest,
    db: Session = Depends(get_db),
    session: UserSession = Depends(require_permission(POLICIES["tickets"].actions["edit"])),
) -> TicketNoteRead:
    """Add an internal ticket note."""
    note = ticketing_service.add_ticket_note(
        db,
        org_id=session.org_id,
        actor_user_id=session.user_id,
        ticket_id=ticket_id,
        body_markdown=data.body_markdown,
    )
    return TicketNoteRead(
        id=note.id,
        author_user_id=note.author_user_id,
        body_markdown=note.body_markdown,
        body_html_sanitized=note.body_html_sanitized,
        created_at=note.created_at,
        updated_at=note.updated_at,
    )


@router.post(
    "/{ticket_id}/link-surrogate",
    response_model=TicketListItem,
    dependencies=[Depends(require_csrf_header)],
)
def link_ticket_surrogate(
    ticket_id: UUID,
    data: TicketLinkSurrogateRequest,
    db: Session = Depends(get_db),
    session: UserSession = Depends(
        require_permission(POLICIES["tickets"].actions["link_surrogates"])
    ),
) -> TicketListItem:
    """Manually link/unlink surrogate mapping for a ticket."""
    ticket = ticketing_service.link_ticket_surrogate(
        db,
        org_id=session.org_id,
        actor_user_id=session.user_id,
        ticket_id=ticket_id,
        surrogate_id=data.surrogate_id,
        reason=data.reason,
    )
    return TicketListItem(
        id=ticket.id,
        ticket_code=ticket.ticket_code,
        status=ticket.status.value if hasattr(ticket.status, "value") else str(ticket.status),
        priority=ticket.priority.value if hasattr(ticket.priority, "value") else str(ticket.priority),
        subject=ticket.subject,
        requester_email=ticket.requester_email,
        requester_name=ticket.requester_name,
        assignee_user_id=ticket.assignee_user_id,
        assignee_queue_id=ticket.assignee_queue_id,
        surrogate_id=ticket.surrogate_id,
        surrogate_link_status=ticket.surrogate_link_status.value
        if hasattr(ticket.surrogate_link_status, "value")
        else str(ticket.surrogate_link_status),
        first_message_at=ticket.first_message_at,
        last_message_at=ticket.last_message_at,
        last_activity_at=ticket.last_activity_at,
        created_at=ticket.created_at,
        updated_at=ticket.updated_at,
    )


@router.get("/{ticket_id}/attachments/{attachment_id}/download")
def download_ticket_attachment(
    ticket_id: UUID,
    attachment_id: UUID,
    db: Session = Depends(get_db),
    session: UserSession = Depends(get_current_session),
) -> dict[str, str]:
    """Return secure signed URL for ticket attachment download."""
    url = ticketing_service.get_ticket_attachment_download_url(
        db,
        org_id=session.org_id,
        actor_user_id=session.user_id,
        ticket_id=ticket_id,
        attachment_id=attachment_id,
    )
    return {"download_url": url}

File: apps/web/lib/api/tickets.ts
/**
 * Ticketing API client.
 */

import api from './index'

export type TicketStatus = 'new' | 'open' | 'pending' | 'resolved' | 'closed' | 'spam'
export type TicketPriority = 'low' | 'normal' | 'high' | 'urgent'
export type TicketLinkStatus = 'unlinked' | 'linked' | 'needs_review'

export interface TicketListItem {
    id: string
    ticket_code: string
    status: TicketStatus
    priority: TicketPriority
    subject: string | null
    requester_email: string | null
    requester_name: string | null
    assignee_user_id: string | null
    assignee_queue_id: string | null
    surrogate_id: string | null
    surrogate_link_status: TicketLinkStatus
    first_message_at: string | null
    last_message_at: string | null
    last_activity_at: string | null
    created_at: string
    updated_at: string
}

export interface TicketMessageAttachment {
    id: string
    attachment_id: string
    filename: string | null
    content_type: string | null
    size_bytes: number
    is_inline: boolean
    content_id: string | null
}

export interface TicketMessageOccurrence {
    id: string
    mailbox_id: string
    gmail_message_id: string
    gmail_thread_id: string | null
    state: string
    original_recipient: string | null
    original_recipient_source: string | null
    original_recipient_confidence: string | null
    original_recipient_evidence: Record<string, unknown>
    parse_error: string | null
    stitch_error: string | null
    link_error: string | null
    created_at: string
}

export interface TicketMessage {
    id: string
    message_id: string
    direction: 'inbound' | 'outbound'
    stitched_at: string
    stitch_reason: string
    stitch_confidence: string
    rfc_message_id: string | null
    gmail_thread_id: string | null
    date_header: string | null
    subject: string | null
    from_email: string | null
    from_name: string | null
    to_emails: string[]
    cc_emails: string[]
    reply_to_emails: string[]
    snippet: string | null
    body_text: string | null
    body_html_sanitized: string | null
    attachments: TicketMessageAttachment[]
    occurrences: TicketMessageOccurrence[]
}

export interface TicketEvent {
    id: string
    actor_user_id: string | null
    event_type: string
    event_data: Record<string, unknown>
    created_at: string
}

export interface TicketNote {
    id: string
    author_user_id: string | null
    body_markdown: string
    body_html_sanitized: string | null
    created_at: string
    updated_at: string
}

export interface TicketSurrogateCandidate {
    id: string
    surrogate_id: string
    confidence: string
    evidence_json: Record<string, unknown>
    is_selected: boolean
    created_at: string
}

export interface TicketDetailResponse {
    ticket: TicketListItem
    messages: TicketMessage[]
    events: TicketEvent[]
    notes: TicketNote[]
    candidates: TicketSurrogateCandidate[]
}

export interface TicketListResponse {
    items: TicketListItem[]
    next_cursor: string | null
}

export interface TicketListParams {
    limit?: number
    cursor?: string
    status?: TicketStatus
    priority?: TicketPriority
    queue_id?: string
    assignee_user_id?: string
    surrogate_id?: string
    needs_review?: boolean
    q?: string
}

export interface TicketPatchRequest {
    status?: TicketStatus
    priority?: TicketPriority
    assignee_user_id?: string | null
    assignee_queue_id?: string | null
}

export interface TicketReplyRequest {
    to_emails: string[]
    cc_emails?: string[]
    subject?: string
    body_text: string
    body_html?: string
    idempotency_key?: string
}

export interface TicketComposeRequest {
    to_emails: string[]
    cc_emails?: string[]
    subject: string
    body_text: string
    body_html?: string
    surrogate_id?: string
    queue_id?: string
    idempotency_key?: string
}

export interface TicketSendResult {
    status: string
    ticket_id: string
    message_id: string
    provider: string
    gmail_message_id: string | null
    gmail_thread_id: string | null
    job_id?: string | null
}

export interface TicketSendIdentity {
    integration_id: string
    account_email: string
    provider: string
    is_default: boolean
}

export interface TicketSendIdentityResponse {
    items: TicketSendIdentity[]
}

function buildQuery(params: TicketListParams = {}): string {
    const search = new URLSearchParams()

    if (params.limit) search.set('limit', String(params.limit))
    if (params.cursor) search.set('cursor', params.cursor)
    if (params.status) search.set('status', params.status)
    if (params.priority) search.set('priority', params.priority)
    if (params.queue_id) search.set('queue_id', params.queue_id)
    if (params.assignee_user_id) search.set('assignee_user_id', params.assignee_user_id)
    if (params.surrogate_id) search.set('surrogate_id', params.surrogate_id)
    if (params.needs_review !== undefined) search.set('needs_review', String(params.needs_review))
    if (params.q) search.set('q', params.q)

    const query = search.toString()
    return query ? `?${query}` : ''
}

export function getTickets(params: TicketListParams = {}): Promise<TicketListResponse> {
    return api.get<TicketListResponse>(`/tickets${buildQuery(params)}`)
}

export function getTicket(ticketId: string): Promise<TicketDetailResponse> {
    return api.get<TicketDetailResponse>(`/tickets/${ticketId}`)
}

export function patchTicket(ticketId: string, data: TicketPatchRequest): Promise<TicketListItem> {
    return api.patch<TicketListItem>(`/tickets/${ticketId}`, data)
}

export function replyTicket(ticketId: string, data: TicketReplyRequest): Promise<TicketSendResult> {
    return api.post<TicketSendResult>(`/tickets/${ticketId}/reply`, data)
}

export function composeTicket(data: TicketComposeRequest): Promise<TicketSendResult> {
    return api.post<TicketSendResult>('/tickets/compose', data)
}

export function addTicketNote(ticketId: string, bodyMarkdown: string): Promise<TicketNote> {
    return api.post<TicketNote>(`/tickets/${ticketId}/notes`, { body_markdown: bodyMarkdown })
}

export function linkTicketSurrogate(
    ticketId: string,
    data: { surrogate_id?: string | null; reason?: string }
): Promise<TicketListItem> {
    return api.post<TicketListItem>(`/tickets/${ticketId}/link-surrogate`, data)
}

export function getTicketSendIdentities(): Promise<TicketSendIdentityResponse> {
    return api.get<TicketSendIdentityResponse>('/tickets/send-identities')
}

export async function getTicketAttachmentDownloadUrl(
    ticketId: string,
    attachmentId: string
): Promise<string> {
    const payload = await api.get<{ download_url: string }>(
        `/tickets/${ticketId}/attachments/${attachmentId}/download`
    )
    return payload.download_url
}
